src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -1384,5 +1384,90 @@ void M(bool b)
 
             AssertEx.Equal(expectedTypes, invocationTypes);
         }
+
+        [Fact]
+        public void FunctionPointerTypeCannotBeUsedInDynamicTypeArguments()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(dynamic d)
+    {
+        d.M<delegate*<void>>();
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,13): error CS0306: The type 'delegate*<void>' may not be used as a type argument
+                //         d.M<delegate*<void>>();
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "delegate*<void>").WithArguments("delegate*<void>").WithLocation(6, 13)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeCannotBeUsedInDynamicArgument()
+        {
+
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(dynamic d, delegate*<void> ptr)
+    {
+        d.M(ptr);
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,13): error CS1978: Cannot use an expression of type 'delegate*<void>' as an argument to a dynamically dispatched operation.
+                //         d.M(ptr);
+                Diagnostic(ErrorCode.ERR_BadDynamicMethodArg, "ptr").WithArguments("delegate*<void>").WithLocation(6, 13)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeCannotBeConvertedFromDynamic()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(delegate*<void> ptr)
+    {
+        dynamic d = ptr;
+        ptr = d;
+        ptr = (delegate*<void>)d;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,21): error CS0029: Cannot implicitly convert type 'delegate*<void>' to 'dynamic'
+                //         dynamic d = ptr;
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "ptr").WithArguments("delegate*<void>", "dynamic").WithLocation(6, 21),
+                // (7,15): error CS0029: Cannot implicitly convert type 'dynamic' to 'delegate*<void>'
+                //         ptr = d;
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "d").WithArguments("dynamic", "delegate*<void>").WithLocation(7, 15),
+                // (8,15): error CS0030: Cannot convert type 'dynamic' to 'delegate*<void>'
+                //         ptr = (delegate*<void>)d;
+                Diagnostic(ErrorCode.ERR_NoExplicitConv, "(delegate*<void>)d").WithArguments("dynamic", "delegate*<void>").WithLocation(8, 15)
+            );
+        }
+
+        [Fact]
+        public void FunctionPointerTypeCannotBeQuestionDotted()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M(delegate*<void> ptr)
+    {
+        ptr?.ToString();
+    }
+}");
+
+            comp.VerifyDiagnostics(
```

Comment:
>comp.VerifyDiagnostics( [](start = 12, length = 23)

It looks like implementation change is about inferring result type of the conditional access. The test doesn't really observe the effect of the change. Consider either adding a test for a different scenario where it would be apparent from the error that the type we inferred is void (```_ = ptr?.ToString();```?), or asserting that type through other means (```GetTypeInfo```?);

Reply:
It looks like ```UnusedFunctionPointerAsResultOfQuestionDot``` still doesn't test scenario that changes in ```BindConditionalAccessExpression``` are affecting. I believe the target scenario is when the RHS of a conditional access returns a function pointer.  ```UnusedFunctionPointerAsResultOfQuestionDot``` tests scenarios when RHS returns void.

---
In reply to: [407711198](https://github.com/dotnet/roslyn/pull/43248#discussion_r407711198) [](ancestors = 407711198)

Reply:
Still not testing that the type we infer for the conditional access is void

---
In reply to: [408273921](https://github.com/dotnet/roslyn/pull/43248#discussion_r408273921) [](ancestors = 408273921,407711198)

Reply:
Missed the GetTypeInfo asserts below IL verification

---
In reply to: [408322095](https://github.com/dotnet/roslyn/pull/43248#discussion_r408322095) [](ancestors = 408322095,408273921,407711198)